src/main/java/de/unistuttgart/bugfinder/gameresult/GameResultService.java

```diff
@@ -0,0 +1,200 @@
+        if (solution.get().getBugs().size() == 0 && bugsMissolved == 0) {
+          score = MAX_SCORE;
+        } else {
+          score = (Math.max(bugsSolved - bugsMissolved, 0) / Math.max(solution.get().getBugs().size(), 1)) * 100;
+        }
+        log.debug("Solved {}/{} bugs", bugsSolved, solution.get().getBugs().size());
+        log.debug("Missolved {} not bugs", bugsMissolved);
+        log.debug("Reached score of {} on code wit id {}", score, code.getId());
+        return score;
+      }
+    }
+    return 0l;
+  }
+
+  /**
+   * Returns a list of bugs a player submitted which are according to the solutions no real bugs.
+   *
+   * @param solution the solution of the code entered by a lecturer
+   * @param playerSolution the solution from the player
+   * @return a list of bugs that are not bugs in the solution
+   */
+  private List<BugDTO> getFalseSubmittedBugs(Solution solution, SolutionDTO playerSolution) {
+    return playerSolution
+      .getBugs()
+      .parallelStream()
+      .filter(playerBug ->
+        solution
+          .getBugs()
+          .parallelStream()
+          .filter(bug -> bug.getWord().getId().equals(playerBug.getWordId()))
+          .count() ==
+        0
+      )
+      .toList();
+  }
+
+  /**
+   * Searches for the player's solution of a code.
+   *
+   * @param submittedSolutions the list of submitted solutions
+   * @param code the code to search for the solution
+   * @return an optional of a submitted solution from the player from a specific code
+   */
+  private Optional<SubmittedSolutionDTO> getSubmittedSolutionFromCode(
+    List<SubmittedSolutionDTO> submittedSolutions,
+    Code code
+  ) {
+    return submittedSolutions
+      .parallelStream()
+      .filter(submittedSolution -> submittedSolution.getCodeId().equals(code.getId()))
+      .findAny();
+  }
+
+  /**
+   * Calculate if a bug was fixed or not
+   *
+   * @param bug The bug with correct values
+   * @param playersSolution the players solution
+   * @return whether the bug was fixed or not
+   */
+  private boolean fixedBug(final Bug bug, final SolutionDTO playersSolution) {
+    final Optional<BugDTO> playersBug = playersSolution
+      .getBugs()
+      .parallelStream()
+      .filter(toFindPlayerBug -> toFindPlayerBug.getWordId().equals(bug.getWord().getId()))
+      .findAny();
+    if (playersBug.isEmpty()) {
+      return false;
+    }
+    return (
+      playersBug.get().getErrorType() == bug.getErrorType() &&
+      playersBug.get().getCorrectValue().equals(bug.getCorrectValue())
+    );
+  }
+
+  /**
+   * Calculates the average score of a list of codes.
+   *
+   * @param codeScore a map with score for each code
+   * @return the average score of all code scores
+   */
+  private long calculateScoreFromCodeScores(final Map<Code, Long> codeScore) {
+    long totalScore = 0;
+    for (final long score : codeScore.values()) {
+      totalScore += score;
+    }
+    return totalScore / codeScore.size();
```

Comment:
```suggestion
    return map.values().parallelStream().mapToLong(l -> l).filter(score -> score <= MAX_SCORE && score >= 0).average().orElse(MAX_SCORE);
```
Handles even the edge cases that there is no question, or invalid values, correctly…